### PR TITLE
mon/OSDMonitor: warn on unsafe pool [min] size

### PR DESCRIPTION
**Pull request by xiexingguo**
Replace https://github.com/ceph/ceph/pull/13808

Signed-off-by: xie xingguo <xie.xingguo@zte.com.cn>

### CODE REVIEW

**Comment by liewegas** [comment]
These health checks are all about to change; do you mind waiting for #15643 ?

**Comment by xiexingguo** [comment]
> These health checks are all about to change; do you mind waiting for #15643 ?

will rebase once it gets merged.

**Comment by liewegas** [comment]
#15643 is merged, rebase?

**Comment by xiexingguo** [comment]
retest this please

**Comment by xiexingguo** [comment]
retest this please

**Comment by xiexingguo** [comment]
@liewegas Do we need to this one into Luminous?

**Comment by liewegas** [comment]
@xiexingguo mind rebasing this on master (notably, adding this to options.cc)?

File: src/osd/OSDMap.cc [hunk 1]
```diff
@@ -4265,6 +4267,43 @@ void print_osd_utilization(const OSDMap& osdmap,
...
+      if (plugin == profile.end()) {
+        break;
+      }
+      ErasureCodePluginRegistry &instance =
+        ErasureCodePluginRegistry::instance();
+      int err = instance.factory(plugin->second,
+                                 g_conf->get_val<std::string>("erasure_code_dir"),
+                                 profile,
+                                 &erasure_code,
+                                 &ss);
```
Comment (liewegas) [inline src/osd/OSDMap.cc]: I'm not wild about loading up the plugin every time we're doing this health check.  Maybe we can cache it somewhere?

File: src/osd/OSDMap.cc [hunk 2]
```diff
@@ -4605,21 +4644,71 @@ void OSDMap::check_health(health_check_map_t *checks) const
 
   // POOL_FULL
   {
-    list<string> detail;
+    enum {
+      WARN_POOL_FULL,
+      WARN_POOL_SIZE,
+      WARN_POOL_MIN_SIZE,
+      WARN_POOL_MAX_TYPE,
```
Comment (liewegas) [inline src/osd/OSDMap.cc]: Is this enum helpful?  We're only ever referencing vector positions explicitly; we don't programatically loop over them.  May as well just do
<pre>
list<string> detail_full, detail_size, detail_min_size, detail_max_type;
</pre>
?

File: src/common/config_opts.h [hunk 3]
```diff
@@ -332,6 +332,11 @@ OPTION(mon_warn_on_crush_straw_calc_version_zero, OPT_BOOL, true) // warn if cru
 OPTION(mon_warn_on_osd_down_out_interval_zero, OPT_BOOL, true) // warn if 'mon_osd_down_out_interval == 0'
 OPTION(mon_warn_on_cache_pools_without_hit_sets, OPT_BOOL, true)
 OPTION(mon_warn_osd_usage_min_max_delta, OPT_FLOAT, .40) // warn if difference between min and max OSD utilizations exceeds specified amount
+// note that for ec, the minimal sensible configuration is 2 + 1,
+// so 3 is applicable for both replicated pools and ec pools
+OPTION(mon_suggested_pool_size, OPT_U32, 3)
```
Comment (liewegas) [inline src/common/config_opts.h]: It's convenient that 3 is what we (i.e., developers) want for both ec and replicated, but I suspect that that may be vary in other cases. Perhaps we want a mon_suggested_replicas and mon_suggested_m (as in k+m) or something?

**Comment by xiexingguo** [comment]
Will refactor and repush

**Pull request closed without merging**